substrate/primitives/consensus/beefy/src/lib.rs

```diff
@@ -352,6 +383,55 @@ where
 	return valid_first && valid_second
 }
 
+/// Validates [ForkEquivocationProof] by checking:
+/// 1. `commitment` is signed,
+/// 2. `correct_header` is valid and matches `commitment.block_number`.
+/// 2. `commitment.payload` != `expected_payload(correct_header)`.
+/// NOTE: GRANDPA finalization proof is not checked, which leads to slashing on forks.
+/// This is fine since honest validators will not be slashed on the chain finalized
+/// by GRANDPA, which is the only chain that ultimately matters.
+/// The only material difference not checking GRANDPA proofs makes is that validators
+/// are not slashed for signing BEEFY commitments prior to the blocks committed to being
+/// finalized by GRANDPA. This is fine too, since the slashing risk of committing to
+/// an incorrect block implies validators will only sign blocks they *know* will be
+/// finalized by GRANDPA.
+pub fn check_fork_equivocation_proof<Number, Id, MsgHash, Header>(
+	proof: &ForkEquivocationProof<Number, Id, <Id as RuntimeAppPublic>::Signature, Header>,
+	expected_header_hash: &Header::Hash,
+) -> bool
+where
+	Id: BeefyAuthorityId<MsgHash> + PartialEq,
+	Number: Clone + Encode + PartialEq,
+	MsgHash: Hash,
+	Header: HeaderT,
+{
+	let ForkEquivocationProof { commitment, signatories, correct_header } = proof;
+
+	if correct_header.hash() != *expected_header_hash {
+		return false
+	}
+
+	let expected_mmr_root_digest = mmr::find_mmr_root_digest::<Header>(correct_header);
+	let expected_payload = expected_mmr_root_digest
+		.map(|mmr_root| Payload::from_single_entry(known_payloads::MMR_ROOT_ID, mmr_root.encode()));
+
+	// cheap failfasts:
+	// 1. check that `payload` on the `vote` is different that the `expected_payload`
+	// 2. if the signatories signed a payload when there should be none (for
+	// instance for a block prior to BEEFY activation), then expected_payload =
+	// None, and they will likewise be slashed
+	if Some(&commitment.payload) != expected_payload.as_ref() {
+		// check check each signatory's signature on the commitment.
+		// if any are invalid, equivocation report is invalid
+		// TODO: refactor check_commitment_signature to take a slice of signatories
+		return signatories.iter().all(|(authority_id, signature)| {
```

Comment:
good observation! I thought I was already filtering for this at the call site - added in https://github.com/paritytech/polkadot-sdk/pull/1903/commits/b19c0923660d5af2e9dbfcf0a3bdf27d165b7767.

My current preference is to leave the responsibility for ensuring correct equivocation reports at the call site (i.e. the fisherman). We can also add this here, but if we are permissive in the runtime processing these, then we have to worry more about & increase our catering for DoS vectors like you point out [here](https://github.com/paritytech/polkadot-sdk/pull/1329#discussion_r1417060121).
wdyt?

Reply:
> My current preference is to leave the responsibility for ensuring correct equivocation reports at the call site (i.e. the fisherman)

I don't think we can leave this check only in the fisherman. We have to check this in the pallet as well, cause otherwise we might receive equivocation reports from other callers (for example someone manually executing an extrinsic) and we would risk not validating them.

Reply:
But I think this is fixed. Please move the conversation to https://github.com/paritytech/polkadot-sdk/pull/1903 if it isn't